api.php

```diff
@@ -0,0 +1,773 @@
+
+	/**
+	 * Initialize an ApiEntry.
+	 * @param string $route - API entry name.
+	 * @param string $type - API entry type ('GET' or 'POST').
+	 * @param string $databasePermissionField - name of corresponding permission field in database table `wD_ApiPermissions`.
+	 * @param array $requirements - array of API entry parameters names.
+	 * @throws Exception - if invalid type or if requirements is not an array.
+	 */
+	public function __construct($route, $type, $databasePermissionField, $requirements) {
+		if (!in_array($type, array('GET', 'POST', 'JSON')))
+			throw new ServerInternalException('Invalid API entry type');
+		if (!is_array($requirements))
+			throw new ServerInternalException('API entry field names must be an array.');
+		$this->route = cleanRoute($route);
+		$this->type = $type;
+		$this->databasePermissionField = $databasePermissionField;
+		$this->requirements = $requirements;
+	}
+
+	/**
+	 * Return API entry name.
+	 * @return string
+	 */
+	public function getRoute() {
+		return $this->route;
+	}
+
+	/**
+	 * Return API entry permission field name.
+	 * @return string
+	 */
+	public function getPermissionField() {
+		return $this->databasePermissionField;
+	}
+
+	/**
+	 * Return an array of actual API parameters values, retrieved from $_GET or $_POST, depending on API entry type.
+	 * @return array
+	 * @throws RequestException
+	 */
+	public function getArgs() {
+		$rawArgs = array();
+		if ($this->type == 'GET')
+			$rawArgs = $_GET;
+		else if ($this->type == 'POST')
+			$rawArgs = $_POST;
+		else if ($this->type == 'JSON') {
+			$rawArgs = json_decode(file_get_contents("php://input"), true);
+			if (!$rawArgs)
+				throw new RequestException('Invalid JSON request data.');
+		}
+		$selectedArgs = array();
+		foreach ($this->requirements as $fieldName) {
+			$selectedArgs[$fieldName] = isset($rawArgs[$fieldName]) ? $rawArgs[$fieldName] : null;
+		}
+		return $selectedArgs;
+	}
+
+	/**
+	 * Return true if this API entry requires a parameter called `gameID`.
+	 */
+	public function requiresGameID() {
+		return in_array('gameID', $this->requirements);
+	}
+
+	/**
+	 * Return Game object for game associated to this API entry call.
+	 * To get associated game, API entry must expect a parameter named `gameID`.
+	 * @return Game
+	 * @throws RequestException - if no gameID field in requirements, or if no valid game ID provided.
+	 */
+	public function getAssociatedGame() {
+		global $DB;
+		if (!in_array('gameID', $this->requirements))
+			throw new RequestException('No game ID available for this request.');
+		$args = $this->getArgs();
+		$gameID = $args['gameID'];
+		if ($gameID == null)
+			throw new RequestException('Game ID not provided.');
+		$gameID = intval($gameID);
+		$Variant = libVariant::loadFromGameID($gameID);
+		libVariant::setGlobals($Variant);
+		$gameRow = $DB->sql_hash('SELECT * from wD_Games WHERE id = '.$gameID);
+		if (!$gameRow)
+			throw new RequestException('Invalid game ID');
+		return new Game($gameRow);
```

Comment:
Shouldn't this return $Variant->Game() or some such? I remember that you need to request all new objects from the variant layer to give the variant a chance to modify the game.

Reply:
`game.php` seems to set `$this->Variant` using the value of `$GLOBALS['Variant']` which is already set at L250.

```
		$gameID = intval($gameID);
		$Variant = libVariant::loadFromGameID($gameID);
		libVariant::setGlobals($Variant);
```

Do you think that should work, or is there a specific method I need to call?